sematic/api/endpoints/runs.py

```diff
@@ -179,6 +198,114 @@ def get_run_endpoint(user: Optional[User], run_id: str) -> flask.Response:
     return flask.jsonify(payload)
 
 
+@sematic_api.route("/api/v1/runs/<run_id>/schedule", methods=["POST"])
+@authenticate
+def schedule_run_endpoint(user: Optional[User], run_id: str) -> flask.Response:
+    """Schedule the run for execution on external compute, like k8s."""
+    try:
+        run = get_run(run_id)
+    except NoResultFound:
+        return jsonify_error(
+            "No runs with id {}".format(repr(run_id)), HTTPStatus.NOT_FOUND
+        )
+
+    n_existing_jobs = len(run.external_jobs)
+    resolution = get_resolution(run.root_id)
+    run = schedule_run(run, resolution)
+    n_jobs_post_schedule = len(run.external_jobs)
+    if n_jobs_post_schedule <= n_existing_jobs:
+        return jsonify_error(
+            "Failed to schedule the run {}, no external jobs".format(repr(run_id)),
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+        )
+    logger.info("Scheduled run with external job: %s", run.external_jobs[-1])
+    if run.future_state != FutureState.SCHEDULED.value:
+        return jsonify_error(
+            "Failed to schedule the run {}".format(repr(run_id)),
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+        )
+    run.started_at = datetime.datetime.utcnow()
+    save_run(run)
+    payload = dict(
+        content=run.to_json_encodable(),
+    )
+    return flask.jsonify(payload)
+
+
+@sematic_api.route("/api/v1/runs/future_states", methods=["POST"])
+@authenticate
+@retry(_DetectedRunRaceCondition, tries=3, delay=10, jitter=1)
+def update_run_status_endpoint(user: Optional[User]) -> flask.Response:
+    """Update the state of runs based on external job status, and return results"""
+    input_payload: Dict[str, Any] = flask.request.json  # type: ignore
+    if "run_ids" not in input_payload:
+        return jsonify_error(
+            "Call did not contain json with a 'run_ids' key", HTTPStatus.BAD_REQUEST
+        )
+    run_ids = input_payload["run_ids"]
+
+    db_status_dict = get_run_status_details(run_ids)
+    missing_run_ids = set(run_ids).difference(db_status_dict.keys())
+    if len(missing_run_ids) != 0:
+        return jsonify_error(
+            f"Missing runs with ids: {','.join(missing_run_ids)}", HTTPStatus.NOT_FOUND
+        )
+
+    result_list = []
+    for run_id, (future_state, jobs) in db_status_dict.items():
+        new_future_state, message, new_external_jobs = update_run_status(
+            future_state, jobs
+        )
+        run_modified = False
+        run = None
+        if new_external_jobs != jobs:
+            run = get_run(run_id)
+            run.external_jobs = new_external_jobs
+            logger.info("Updating run's external jobs: %s", new_external_jobs)
+            run_modified = True
+        if new_future_state != future_state:
+            run = get_run(run_id) if run is None else run
```

Comment:
More `run = get_run()` online 260 to avoid this and like 262.

Reply:
the thing is, this query will happen A LOT (it's the poll), and we want to avoid loading the full run from the DB unless we detect there has been a change that requires us to update the run (which should be uncommon relative to the "nothing changed" case). I'll add a comment to clarify